XenModel/PoolJoinRules.cs

```diff
@@ -673,5 +687,51 @@ public static bool HasCompatibleManagementInterface(IXenConnection slaveConnecti
             else
                 return slaveConnection.Cache.PIFs.Any(p => !p.physical && p.management && p.VLAN != -1);
         }
+
+        public static bool HasIpForClusterNetwork(IXenConnection masterConnection, Host slaveHost, out bool clusterHostInBond)
+        {
+
+            var clusterHost = masterConnection.Cache.Cluster_hosts.FirstOrDefault();
+
+            if (clusterHost == null)
+            {
+                clusterHostInBond = false;
+                return true;
+            }
+            var clusterHostPif = clusterHost.Connection.Resolve(clusterHost.PIF);
+            clusterHostInBond = clusterHostPif.IsBondNIC();
+
+            var pifsWithIPAddress = 0;
+
+            List<string> ids = new List<string>();
+
+            if (clusterHostInBond)
+            {
+
+                List<PIF> slaves = new List<PIF>();
+
+                foreach (Bond bond in masterConnection.ResolveAll(clusterHostPif.bond_master_of))

```

Comment:
This code re-resolves the bonds for each iteration (you can test this by putting a break point in the ResolveAll method while iterating). Normally the way to do this is to store the resolved bonds in a local variable:
```
var bonds =  masterConnection.ResolveAll(clusterHostPif.bond_master_of);
foreach (var bond in bonds)... etc.
```
Same for line 725.